src/libs/ck-libs/ampi/ampi.C

```diff
@@ -2227,41 +2248,60 @@ void ampi::splitPhase1(CkReductionMsg *msg) noexcept
       newComm = keys[i].nextSplitComm;
   }
 
+  // Count how many colors there are, which is how many reductions to expect
+  {
+    int numColors = 1;
+    int lastColor = keys[0].color;
+    for (int c = 0; c < nKeys; ++c)
+    {
+      if (keys[c].color != lastColor)
+      {
+        lastColor = keys[c].color;
+        ++numColors;
+      }
+    }
+    CkAssert(numCommCreationsInProgress == 0);
+    numCommCreationsInProgress = numColors;
+  }
+
   //Loop over the sorted keys, which gives us the new arrays:
-  int lastColor=keys[0].color-1; //The color we're building an array for
-  CProxy_ampi lastAmpi; //The array for lastColor
+  int lastColor=keys[0].color; //The color we're building an array for
   int lastRoot=0; //C value for new rank 0 process for latest color
-  ampiCommStruct lastComm; //Communicator info. for latest color
-  for (int c=0;c<nKeys;c++) {
-    if (keys[c].color!=lastColor)
+  for (int c=0;c<nKeys;c++)
+  {
+    if (keys[c].color != lastColor)
     { //Hit a new color-- need to build a new communicator and array
-      lastColor=keys[c].color;
-      lastRoot=c;
-
-      if (c!=0) lastAmpi.doneInserting();
-      lastAmpi = createNewChildAmpiSync();
-
       std::vector<int> indices; //Maps rank to array indices for new array
```

Comment:
We can initalize indices to size 'lastRoot' now, and assign in the loop rather than repeatedly calling push_back